Improve "introduction" page, add "Advanced Features" page

cewing opened the pull request:
This is a work in progress, please do not merge.

refs #2614

cewing commented:
@stevepiercy can you help me figure out what I've done to break the build on this PR?  And if you have time, some feedback on the direction I'm taking with these changes would be deeply appreciated.

stevepiercy commented:
@cewing I checked the build history https://ci.appveyor.com/project/mmerickel/pyramid/history but there is no record of the attempt to build when you made your last commit. I'd say it was just a temporary thing that might resolve itself with your next commit.

bertjwregeer commented:
Looks like something went wrong with Appveyor, it seems to think this is an un-mergeable pull request.

cewing commented:
I'll rebase on the changes from master and hopefully that will clear it up.  As it is a WIP, I'm not too worried for the time being, but it'd be nice not to have too big a mess to clean up once the work is done :)

stevepiercy commented:
@cewing I forgot to mention, I like the direction you are headed. Active headings, simpler phrasing... all good stuff.

tseaver commented:
@cewing Wouldn't rebasing against master be cleaner than a merge which adds 168 commits? ;)

mcdonc commented:
My bad.  I told him to just merge.

tseaver commented:
OK.  I sometimes find commit-by-commit review easier, and so try not to do big merges on "reviewable" branches.

cewing commented:
Don't worry, @tseaver, I'll clean up before I ask anyone to review this.

stevepiercy commented:
@cewing walked me through the steps and cleaned this up.

mmerickel commented:
What's the deal with this PR? @stevepiercy, @cewing is this still active?

stevepiercy commented:
@mmerickel rather than close this issue, I'd prefer to hear from @cewing, say by Apr 14. If he doesn't respond, then I'll take up the task. I didn't want to take over his work thus far without giving him a deadline. It's good work, and it would be a shame to close it and forget about it.

mmerickel commented:
That's fine but you can appreciate that I already asked 4 months ago with no response. Please close or work on it at some point. Thanks!

cewing commented:
apologies, @mmerickel and @stevepiercy.  I've had a new job since this work started, and had lost track of this work.  I'd like to continue it, but can't commit to any timeline.  Things are a bit hectic.  I'll be at PyCon and planned on sprinting there.  Can it wait until then?

mmerickel commented:
@cewing of course. There's no rush so long as you intend to work on it I'll leave it open. I just didn't want to leave it open when it looked abandoned!

stevepiercy commented:
@cewing good to hear. Let's revisit at that time. Although I won't attend PyCon this year, I know that the Pylons Project will be sprinting.

stevepiercy commented:
@cewing please update the conversation when it is ready for final review. Make sure you rebase this PR against master, too. There were recent changes in `conf.py` for pytest's URL, and possibly elsewhere, that are causing the docs not to build in Travis CI.

cewing commented:
@stevepiercy, @mmerickel 

This PR is now ready for review.  In short, here's what I've done:

1.  Changed section headers to make them more active
2.  Updated text to make it easier to digest for new developers
3.  Broke more esoteric features out into a new "Advanced Features" doc.  

I'm happy to make whatever adjustments are requested.

stevepiercy commented:
@cewing we're gonna need a bigger 🍺 . I'll look this over tonight with review comments, if any. Thank you!

stevepiercy commented on docs/narr/introduction.rst:
```diff
@@ -9,126 +9,156 @@
+
+Pyramid provides a set of features that are unique among Python web frameworks.
+Others may provide some, but only Pyramid provides them all, in one place,
+fully documented, and useful *à la carte* without needing to pay for the whole
```
and *à la carte* without paying for the whole

stevepiercy commented on docs/narr/introduction.rst:
```diff
@@ -139,103 +169,93 @@ decorators to localize the configuration.  For example:
-configuration instead of decorators to add views. Pyramid decorators are inert
-instead of eager. You detect and activate them with a :term:`scan`.
+However, using Pyramid configuration decorators does not change your code. It
+remains easy to extend, test or reuse. You can test your code as if the
```
extend, test, or reuse. Oxford comma FTW!

stevepiercy commented on docs/narr/introduction.rst:
```diff
@@ -139,103 +169,93 @@ decorators to localize the configuration.  For example:
+can use ``proutes`` and ``pviews`` to inspect how URLs are connected to your
+application code.
 
 Examples: :ref:`debug_authorization_section` and :ref:`command_line_chapter`.
```
Add :doc:`pscripts/index` (I think that will work). We did not have these auto-documented when this was first written.

stevepiercy commented on docs/narr/introduction.rst:
```diff
@@ -139,103 +169,93 @@ decorators to localize the configuration.  For example:
+
+Pyramid add-ons extend the core of the framework with useful abilities. There
+are add-ons available for your favorite template language, SQL and NoSQL
+databases, authentication services and much much more.
```
Too much much.

stevepiercy commented on docs/narr/introduction.rst:
```diff
@@ -275,62 +295,63 @@ Here's a few views defined as methods of a class instead:
-favorite templating system.  You'll then be able to use that templating system
-from within Pyramid just as you'd use one of the "built-in" templating systems.
+In Pyramid, the job of creating a ``Response`` belongs to a :term:`renderer`.
+Any templating system--Mako, Genshi, Chameleon, Jinja2--can be a renderer. In
```
Use em-dash instead of two hyphens:

```
—
```

Also please ask @fangpenlin about Genshi's support for latest version of Pyramid (they were sprinting, and they maintain the package at https://github.com/fangpenlin/pyramid_genshi). If it's no longer supported, then let's remove it from the list. If it is supported, then encourage them to submit an issue to [add an item to Extending Pyramid on trypyramid.com](https://github.com/Pylons/trypyramid.com/issues/202).

fangpenlin replied:
I just tried to test Genshi with latest pyramid and python 3.6, turns out Genshi doesn't work with some syntax since 3.4 since ast module changes

https://travis-ci.org/fangpenlin/pyramid_genshi/jobs/235551465#L350

and as Genshi is not updated for a really long while, I don't see too much people using it, even myself is not using it that much, so I am thinking about dropping support of it. You can remove the Genshi support from list now, if there is really something changes, you can add it back later.

cewing replied:
Thanks for the update @fangpenlin.  I'll remove the claim of support for now.  Let's see if anyone picks this up.

stevepiercy commented on docs/narr/introduction.rst:
```diff
@@ -275,62 +295,63 @@ Here's a few views defined as methods of a class instead:
-returns "the right stuff" in the dictionary.  You can write "real" unit tests
-instead of functionally testing all of your views.
+When you use a :term:`renderer` with your view callable, you are freed from
+needing to return a "webby" ``Response`` object. Instead, your views can return
```
Instead your

stevepiercy commented on docs/narr/introduction.rst:
```diff
@@ -353,38 +374,46 @@ You can return a Python dictionary:
 
-Event system
-~~~~~~~~~~~~
+Use events to coordinate
```
"Use events to coordinate actions". "actions" seems to fit, but there may be a better word or term.

stevepiercy commented on docs/narr/introduction.rst:
```diff
@@ -394,506 +423,80 @@ without conflict with other translations in other domains.
-can just tell Pyramid to configure your view with an ``http_cache`` statement,
-and it will take care of the rest::
+Pyramid provides an easy way to *cache* the results of slow or expensive views.
+You can indicate in view configuration that you want a view to be cached::
```
use ``.. code-block:: python`` You might need to remove the trailing dots after the ``def()``.

stevepiercy commented on docs/narr/introduction.rst:
```diff
@@ -394,506 +423,80 @@ without conflict with other translations in other domains.
-
-.. code-block:: python
-    :linenos:
+Mistakes happen. Problems crop up. No-one writes bug-free code. Pyramid
```
No one writes

stevepiercy commented on docs/narr/introduction.rst:
```diff
@@ -394,506 +423,80 @@ without conflict with other translations in other domains.
+Mistakes happen. Problems crop up. No-one writes bug-free code. Pyramid
+provides a way to handle the exceptions your code encounters. An
+:term:`exception view` is a special kind of view which is automatically called
+when an particular exception type "bubbles up" without being handled by your
```
when a particular

also s/"bubbles up"/arises

stevepiercy commented on docs/narr/introduction.rst:
```diff
@@ -394,506 +423,80 @@ without conflict with other translations in other domains.
+For example, you might register an exception view for the :exc:`Exception`
+exception type, which will catch *all* exceptions, and present a pretty "well,
+this is embarrassing" page.  Or you might choose to register an exception view
+for only certain application-specific exceptions. You can make a one for when a
```
You can make one

stevepiercy commented on docs/narr/advfeatures.rst:
```diff
@@ -0,0 +1,399 @@
+Advanced :app:`Pyramid` Design Features
```
I'd like the filename to be spelled out, "advanced-features". See below, too.

stevepiercy commented on docs/narr/introduction.rst:
```diff
@@ -394,506 +423,80 @@ without conflict with other translations in other domains.
-trolls" or other people who seem to get their rocks off by berating fellow
-users in our various official support channels.  We try to keep it well-lit and
-new-user-friendly.
+   advfeatures
```
"advanced-features"

stevepiercy commented on docs/narr/introduction.rst:
```diff
@@ -925,61 +528,25 @@ November of that year.
-nor discouraging the decision.
+Similar to :term:`Zope`, :app:`Pyramid` applications may easily be extended. If
+you work within the constraints of the framework, you can produce applications
+that can be reused, modified or extended without needing to modify the original
```
reused, modified, or extended (Oxford comma)

cewing commented:
Thanks for the careful reading, @stevepiercy.  Working on fixes now.

cewing commented on docs/narr/advfeatures.rst:
```diff
@@ -0,0 +1,399 @@
+server and failed?
+
+All these problems are symptoms of "mutable global state", also known as
+"import-time side effects" and arise from the use of "singleton" data structures.
```
I've added some of these to the glossary and converted them to `:term:` references

cewing commented on docs/narr/advfeatures.rst:
```diff
@@ -0,0 +1,399 @@
+
+   def aview(request):
+       return "Hello world!"
+
```
Remember, the first indent is actually not a code indent, it's there to make this code part of the `.. code-block::` directive above.  You are right that there are three spaces there.  It's intended to be so.  In the rendered document that first line is up against the left margin.

stevepiercy replied:
Both the first indent of the code sample and the options (`:linenos:`, etc.) to `.. code-block::` must be at least 1 space. The first indent should be the same for both options and code to avoid unnecessary spacing when code is rendered. Those are the only technical requirements.

The first indent should be the same for all code samples, but I noticed that they are not. For example, in docs/narr/advanced-features.rst, lines 32-37, there are 4 spaces on the first indent. In other places, the first indent is 3. This sets off my OCD alarm.

It would be nice if all code blocks use the same number of spaces for the first indent. Also it would be convenient to use 4 spaces because editors tend to insert 4 spaces for indentation via keyboard shortcuts, making it easier to maintain.

cewing commented on docs/narr/advfeatures.rst:
```diff
@@ -0,0 +1,399 @@
+
+   def anotherview(request):
+       return (403, 'text/plain', "Forbidden")
+
```
Again, this is because the three-space indent is part of the `.. code-block:: directive`.

cewing commented on docs/narr/introduction.rst:
```diff
@@ -9,126 +9,156 @@
+statement coverage [#]_ and 95% decision/condition coverage. [#]_ It is
+automatically tested using `Travis <https://travis-ci.org/Pylons/pyramid>`_ and
+`Jenkins <http://jenkins.pylonsproject.org/job/pyramid/>`_ on Python 2.7,
+Python 3.4, Python 3.5, and PyPy after each commit to its GitHub repository.
```
good plan. will change wording accordingly.  thx

cewing commented:
@stevepiercy, I've completed fixes to the advanced features file per your code review.  I've also renamed the file as you requested, which seems to have marked all the diffs as out-of-date.  You might take a final look at that file before giving it the go-ahead.

I'll finish the intro text CR fixes tomorrow or soon.

stevepiercy commented on docs/narr/introduction.rst:
```diff
@@ -73,44 +73,47 @@ working on a framework that is up-to-date and forward-looking.
+and 95% decision/condition coverage. (as measured by `instrumental <http://instrumental.readthedocs.io/en/latest/intro.html>`_)
+It is automatically tested using `Travis <https://travis-ci.org/Pylons/pyramid>`_
+and `Jenkins <http://jenkins.pylonsproject.org/job/pyramid/>`_
+on supported versions of Python after each commit to its GitHub repository.
```
Although Plone docs use semantic line breaks, that went over like a lead balloon with Pylons Project core developers: https://github.com/Pylons/pyramid/issues/2633

One line per sentence was kinda sorta OK.

One line per paragraph in .rst only is preferred.

Otherwise the changes look good.

mmerickel commented on docs/narr/advanced-features.rst:
```diff
@@ -0,0 +1,423 @@
+
+    if __name__ == '__main__':
+        config = Configurator()
+        config.add_response_adapter(string_response_adapter, basestring)
```
`str`, the docs are focused on python 3 over python 2.

mmerickel commented on docs/narr/advanced-features.rst:
```diff
@@ -0,0 +1,423 @@
+
+    if __name__ == '__main__':
+        config = Configurator()
+        config.add_response_adapter(string_response_adapter, basestring)
```
`str`, the docs are focused on python 3 over python 2.

cewing commented:
Okay.  @mmerickel, @stevepiercy, this should be ready to go, unless you have more requested changes.  I think I've addressed everything from your reviews.  Thanks for the thoughtful and close reading!

c

stevepiercy commented on docs/narr/introduction.rst:
```diff
@@ -353,547 +398,133 @@ You can return a Python dictionary:
-at the top of the screen based on an enumeration of views they registered.
-
-This is possible using Pyramid's :term:`introspector`.
+Mistakes happen. Problems crop up. No-one writes bug-free code. :app:`Pyramid`
```
"No one" is not hyphenated.

stevepiercy commented on docs/narr/introduction.rst:
```diff
@@ -353,547 +398,133 @@ You can return a Python dictionary:
+Mistakes happen. Problems crop up. No-one writes bug-free code. :app:`Pyramid`
+provides a way to handle the exceptions your code encounters. An
+:term:`exception view` is a special kind of view which is automatically called
+when an particular exception type "bubbles up" without being handled by your
```
"exception type arises without"

cewing replied:
oh, and "a particular" :)

stevepiercy replied:
good catch. Misteaks happen. 😛

stevepiercy commented on docs/narr/introduction.rst:
```diff
@@ -353,547 +398,133 @@ You can return a Python dictionary:
+For example, you might register an exception view for the :exc:`Exception`
+exception type, which will catch *all* exceptions, and present a pretty "well,
+this is embarrassing" page.  Or you might choose to register an exception view
+for only certain application-specific exceptions. You can make a one for when a
```
"You can make one"

cewing commented:
Okay, @stevepiercy, fixed those last few, found a couple more of my own, and reflowed both docs to use the one-line-per-paragraph style.

stevepiercy commented on docs/narr/introduction.rst:
```diff
@@ -9,126 +9,98 @@
+Documented
+~~~~~~~~~~
+
+The :app:`Pyramid` documentation is comprehensive. We strive to keep our narrative documentation both complete and friendly to newcomers. We also maintain the :ref:`Pyramid Community Cookbook <cookbook:pyramid-cookbook>` of  ecipes demonstrating common scenarios you might face. Contributions in the form of improvements to our documentation are always appreciated. And we always welcome improvement ...
```
"of  ecipes" is missing an "r".

cewing commented:
heh.  Reflow error.  Fixed now.  Have at it, @mmerickel!  And thanks again for being a great editor, @stevepiercy

mmerickel commented:
Fantastic work @cewing !

stevepiercy commented:
Thank you!

tseaver commented:
FWIW, this PR would have been a really good candidate for a squash merge.
